x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1160,12 +1168,592 @@ private synchronized SharedGroupFactory getNettySharedGroupFactory(Settings sett
-            new SystemIndexDescriptor(SECURITY_MAIN_ALIAS, "Contains Security configuration"),
-            new SystemIndexDescriptor(RestrictedIndicesNames.INTERNAL_SECURITY_MAIN_INDEX_6, "Contains Security configuration"),
-            new SystemIndexDescriptor(RestrictedIndicesNames.INTERNAL_SECURITY_MAIN_INDEX_7, "Contains Security configuration"),
-
-            new SystemIndexDescriptor(RestrictedIndicesNames.SECURITY_TOKENS_ALIAS, "Contains auth token data"),
-            new SystemIndexDescriptor(RestrictedIndicesNames.INTERNAL_SECURITY_TOKENS_INDEX_7, "Contains auth token data")
+            SystemIndexDescriptor.builder()
+                // This can't just be `.security-*` because that would overlap with the tokens index pattern
+                .setIndexPattern(".security-[0-9]+")
+                .setPrimaryIndex(RestrictedIndicesNames.INTERNAL_SECURITY_MAIN_INDEX_7)
+                .setDescription("Contains Security configuration")
+                .setMappings(getIndexMappings())
+                .setSettings(getIndexSettings())
+                .setAliasName(SECURITY_MAIN_ALIAS)
+                .setIndexFormat(INTERNAL_MAIN_INDEX_FORMAT)
+                .setVersionMetaKey("security-version")
+                .setOrigin(SECURITY_ORIGIN)
+                .build(),
+
+            SystemIndexDescriptor.builder()
+                .setIndexPattern(".security-tokens-[0-9]+")
+                .setPrimaryIndex(RestrictedIndicesNames.INTERNAL_SECURITY_TOKENS_INDEX_7)
+                .setDescription("Contains auth token data")
+                .setMappings(getTokenIndexMappings())
+                .setSettings(getTokenIndexSettings())
+                .setAliasName(SECURITY_TOKENS_ALIAS)
+                .setIndexFormat(INTERNAL_TOKENS_INDEX_FORMAT)
+                .setVersionMetaKey(SECURITY_VERSION_STRING)
+                .setOrigin(SECURITY_ORIGIN)
+                .build()
         );
     }
+
+    private Settings getIndexSettings() {
+        return Settings.builder()
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+            .put(IndexMetadata.SETTING_AUTO_EXPAND_REPLICAS, "0-1")
+            .put(IndexMetadata.SETTING_PRIORITY, 1000)
+            .put("index.refresh_interval", "1s")
+            .put(IndexMetadata.INDEX_FORMAT_SETTING.getKey(), INTERNAL_MAIN_INDEX_FORMAT)
+            .put("analysis.filter.email.type", "pattern_capture")
+            .put("analysis.filter.email.preserve_original", true)
+            .putList("analysis.filter.email.patterns", List.of("([^@]+)", "(\\p{L}+)", "(\\d+)", "@(.+)"))
+            .put("analysis.analyzer.email.tokenizer", "uax_url_email")
+            .putList("analysis.analyzer.email.filter", List.of("email", "lowercase", "unique"))
+            .build();
+    }
+
+    private XContentBuilder getIndexMappings() {
+        try {
+            final XContentBuilder builder = jsonBuilder();
```

Comment:
Personally I would prefer the mapping as a resource file, I wonder what's the reason for this change.

Reply:
I was following the example of other plugins that auto-created their indices. It does also make it harder to change the mappings, versus opening up a jar file and editing the json.